Documentation/configuration/metrics.rst

```diff
@@ -29,6 +29,12 @@ Endpoint
 * ``endpoint_regenerating``: Number of endpoints currently regenerating
 * ``endpoint_regenerations``: Count of all endpoint regenerations that have completed, tagged by outcome
 
+Drops/Forwards (L3/L4)
+---------------------
+
+* ``drop_count_total``: Total drop requests, tagged by drop reason and ingress/egress direction
+* ``forward_count_total``: Total forward requests, tagged by ingress/egress direction
```

Comment:
I'm still struggling with this direction. Firstly, similar my comments above, "forward requests" means potentially something different from "forwarded requests", and also from a BPF perspective it handles packets, not requests. For argument's sake, let's say this is actually "Total forwarded packets". 

Even looking at "Total forwarded packets" under the "L3/L4" heading, one might draw the conclusion that if this counter doesn't go up, then the packets were never forwarded in the L3/L4 part of processing. However, what's actually happening is that the packets are going to the proxy. Based on the below description, these are explicitly not counted. Furthermore, the counter is also not incremented when the packet comes back from the proxy. In such a case where an L7 policy applies to some set of traffic, the "L3/L4 forwarded" metric is *never incremented*.

https://github.com/cilium/cilium/blob/7281c7e994707b5721a4fffc00b6ff3032ee9daf/bpf/lib/trace.h#L114-L116

I think that this is worth describing somehow here. Perhaps it's a matter of stating that "If L7 policy applies, then the packets will never be counted against ``forwarded_count_total``." ?

There's still an (albeit small) possibility that packets would not hit an L3/L4 drop metric, not hit the L3/L4 forwarded metric (because it goes to the proxy), then proceed to be dropped before it reaches the proxy, perhaps for instance because the packet had invalid checksums or something which we don't detect in L3/L4, but the stack will handle. In that case, no counters would be incremented at all. I wonder whether we should attempt to provide visibility into this kind of thing. AFAICT with the current approach, such packets would be ghosts - never counted in any metric even though they went through Cilium.

Reply:
Even if L7 is applied and the packet is forwarded into an endpoint, the L3 counter should get bumped eventually. Right now, the following scenarios exist:
1. Drop at L3
2. Forward at L3, no L7
3. Drop at L7, no L3 counter incremented
4. Forward at L7, L3 forward counter is incremented

I've considered suggesting to bump the counters on the way to the proxy but it seems weird to bump them twice if the packets eventually make it to the endpoint. What we could consider is to only bump at TO_LXC if it is **not** coming from the proxy.

Reply:
Oh. I was concerned about the case of ingress proxy towards the endpoint, however I now believe that according to your logic above, the packet would hit the ingress `bpf_lxc` prog, see that L7 must occur, send to the proxy, the proxy allows it, then the packet comes back from the proxy into `bpf_netdev` and gets forwarded back into the `bpf_lxc` prog again, this time with the SKIP_PROXY flag set, then eventually will hit `send_trace_notify()` there (same as it would in a pure L3/L4 allow)  and increment the counter. In summary I believe that this case is fine.

However for packets sent through the egress proxy (case 4 above, but for egress part of the datapath), I don't think that the L3 "forwarded" metric is ever incremented:

https://github.com/cilium/cilium/blob/0136423913772516756e46f2469311684002106d/bpf/bpf_netdev.c#L380

https://github.com/cilium/cilium/blob/0136423913772516756e46f2469311684002106d/bpf/lib/l3.h#L122-L129